### PR TITLE
Revamp persistence layer

### PR DESCRIPTION
**Pull request by smondet**

### CODE REVIEW

**Comment by smondet** [comment]
@ihodes This ready for review.

I prefer getting this merged before doing changes that are less related to the database layer.

**Comment by smondet** [comment]
Here is an example of log output:

```
========================================================================[2]
[2016-10-27-22h45m58s788ms-UTC]
* Module: Persistence
* function: query_nodes
* query: { Protocol.Up_message.V0.time_constraint = `Created_after (42.);
  filter =
  `Or ([`And ([`Has_tag (`Equals ("fails"))]);
         `And ([`Has_tag (`Equals ("secret"))])])
  }
* sql-query: SELECT blob FROM ketrew_main WHERE (creation_date >= $3) AND ((tags @> $2) OR (tags @> $1))
* sql-arguments:
    * "`List ([`Blob (\"secret\")])"
    * "`List ([`Blob (\"fails\")])"
    * "`Timestamp (42.)"
* timing:
    * start: 2016-10-27-22h45m58s786ms-UTC
    * all-targets: 0.00172805786133 s.
    * list-of-ids: 1.28746032715e-05 s.
    * total: 0.0017409324646 s.
* before-filter: 3 nodes
* after-filter: 3 nodes
```

Gives this:

image

**Comment by smondet** [comment]
another one:

```
========================================================================[2]
[2016-10-27-23h17m36s074ms-UTC]
* Module: Persistence
* function: query_nodes
* query: { Protocol.Up_message.V0.time_constraint = `Created_after (42.);
  filter = `Name (`Matches ("^Coclo.*ow$")) }
* sql-query: SELECT blob FROM ketrew_main WHERE (creation_date >= $2) AND ((encode(name, 'escape') :: text) ~ $1 )
* sql-arguments:
    * "`Blob (\"^Coclo.*ow$\")"
    * "`Timestamp (42.)"
* timing:
    * start: 2016-10-27-23h17m36s073ms-UTC
    * all-targets: 0.0011990070343 s.
    * list-of-ids: 0.000253915786743 s.
    * total: 0.00145292282104 s.
* before-filter: 1 nodes
* after-filter: 1 nodes
```


image

**Comment by smondet** [comment]
Gives things like this:

```
postgres=# \d+
List of relations
-[ RECORD 1 ]-----------------
Schema      | public
Name        | ketrew_add_list
Type        | table
Owner       | postgres
Size        | 16 kB
Description | 
-[ RECORD 2 ]-----------------
Schema      | public
Name        | ketrew_kill_list
Type        | table
Owner       | postgres
Size        | 48 kB
Description | 
-[ RECORD 3 ]-----------------
Schema      | public
Name        | ketrew_main
Type        | table
Owner       | postgres
Size        | 128 kB
Description | 

postgres=# \d+ ketrew_main;
                                   Table "public.ketrew_main"
         Column          |       Type       | Modifiers | Storage  | Stats target | Description 
-------------------------+------------------+-----------+----------+--------------+-------------
 id                      | bytea            | not null  | extended |              | 
 name                    | bytea            | not null  | extended |              | 
 tags                    | bytea[]          |           | extended |              | 
 creation_date           | double precision | not null  | plain    |              | 
 blob                    | bytea            | not null  | extended |              | 
 engine_status           | bytea            | not null  | extended |              | 
 last_status_change_date | double precision | not null  | plain    |              | 
 finished_date           | double precision |           | plain    |              | 
 simple_status           | bytea            | not null  | extended |              | 
 really_running          | bytea            | not null  | extended |              | 
 killable                | bytea            | not null  | extended |              | 
 dependency_dead         | bytea            | not null  | extended |              | 
 activated_by_user       | bytea            | not null  | extended |              | 
 killed_from_passive     | bytea            | not null  | extended |              | 
 failed_from_running     | bytea            | not null  | extended |              | 
 failed_from_starting    | bytea            | not null  | extended |              | 
 failed_from_condition   | bytea            | not null  | extended |              | 
```

File: src/lib/engine.ml [hunk 1]
```diff
@@ -37,7 +37,7 @@ include Make_module_error_and_info(struct
 let create configuration =
   Persistent_data.create
     (Configuration.database_parameters configuration)
-    (Configuration.archival_age_threshold configuration)
+    (* (Configuration.archival_age_threshold configuration) *)
```
Comment (ihodes) [inline src/lib/engine.ml]: delete line?

File: src/lib/persistent_data.mli [hunk 2]
```diff
@@ -186,18 +157,59 @@ module Synchronize: sig
...
-          | `Write_file_exn of string * exn ]
-     | `Fetching_node of Error.fetching_node
-     | `Not_a_directory of string
-     | `System of
-          [> `File_info of string
-          | `List_directory of string
-          | `Make_directory of string ] *
-          [> `Exn of exn | `Wrong_access_rights of int ]
-     | `Target of [> `Deserilization of string ] ])
+     [> `Syncronize of
```
Comment (ihodes) [inline src/lib/persistent_data.mli]: syncronize → synchronize

File: src/lib/engine.ml (hunk 1 above)
Comment (smondet) [inline src/lib/engine.ml]: Oops right I left it commented to remind me to https://github.com/hammerlab/ketrew/issues/478

**Comment by ihodes** [comment]
I think I get what's going on here. Thank you for the comments! 

#27 th -> the
#30 actual -> actually (or just rephrase to drop actual)
#50 Persistend_data -> Persistent_data
#1724 could be named "remove_equivalent_targets" (just a suggestion)
#1774 activable -> activatable

LGTM, minor things only. Curious why you ended up with three tables instead of just one?

**Comment by smondet** [comment]
@ihodes 

- `remove_equivalent_targets` → they're not really removed
- https://en.wiktionary.org/wiki/activable < https://en.wiktionary.org/wiki/activatable :)

```
$ curl https://en.wiktionary.org/wiki/activatable| wc -c
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 22809    0 22809    0     0  91675      0 --:--:-- --:--:-- --:--:-- 91971
22809
 $ curl https://en.wiktionary.org/wiki/activable| wc -c
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 26353    0 26353    0     0  98433      0 --:--:-- --:--:-- --:--:-- 98700
26353
```


The 3 "tables" were already there before (we were using Trakeva "collections"):

They allow to "save" work to do (add or kill nodes) and delay it; e.g. when a workflow is submitted it gets saved "verbatim", then we can return `OK` to the HTTP-client, and when the times comes (after an `Engine.step`) we can do the equivalence + safely add them to the main table.

**Comment by ihodes** [comment]
LGTM!

**Pull request merged**